alonzo-ty/src/lib.rs

```diff
@@ -0,0 +1,36 @@
+use std::ops::Deref;
+
+pub trait BaseTy: Clone {
+    type Value: Clone;
+}
+
+#[derive(Clone)]
+pub enum Ty<T: BaseTy> {
+    Base(T),
+    Func(Box<Self>, Box<Self>),
+    Product(Vec<Self>),
+    Sum(Vec<Self>),
+}
+
+#[derive(Clone)]
+pub struct TyNode<I, T: BaseTy> {
```

Comment:
I use a similar but more general pattern in Tao for things like this. Perhaps we can abstract it and use it for storing all kinds of metadata common to `alonzo-lir` and `alonzo-mir`? https://github.com/zesterer/tao/blob/master/src/node.rs

Reply:
That sounds good.